keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,215 @@
+which is a deliberate decision. Go embedding which, once flattened, would define
+multiple unions in the same structure are also forbidden. Also note that we have
+no occurrences of these existing in core-types (AFAIK).
+
+### OpenAPI
+
+OpenAPI v3 already allows a "oneOf" form, which is accepted by CRD validation
+(and will continue to be accepted in the future). That oneOf form will be used
+for validation, but is "on-top" of this proposal (validation is considered as
+not included in this proposal).
+
+New extensions are created in the openapi to describe the behavior:
+- `x-kubernetes-union-discriminator: <discriminator>` is set to the name of the
+  discriminator field, if present,
+- `x-kubernetes-oneof-fields-discriminated: {"<fieldName>":
+  "<discriminatedName>"}` is a map of fields that belong to the union to their
+  discriminated names. The discriminatedValue will typically be set to the name
+  of the Go variable.
+
+These fields belong directly in the object that holds the union.
+
+Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.
+
+As for Go types, we only accept one union per structure, which means there can
+be only one discriminator, and all other `oneof` fields are assumed to belong to
+the same union.
+
+### Discriminator
+
+For backward compatibility reasons, discriminators can be added to existing
+union structures as an optional string. This has a nice property that it's going
+to allow conflict detection when the selected union field is changed.
+
+The value of the discriminator is going to be set automatically by the apiserver
+when a new field is changed in the union. It will be set to the value of the
+`x-kubernetes-union-fields-discriminated` extension for that specific field.
+
+When the value of the discriminator is explicitly changed by the client, it
+will be interpreted as an intention to clear all the other fields. See section
+below.
+
+If the object had exactly one of the union field set, and that field is removed,
+the discriminator will be cleared.
+
+### Normalizing on updates
+
+A "normalization process" will run automatically for all creations and
+modifications (either with put or patch). It will happen automatically in order
+to clear fields and update discriminators. This process will run for both
+core-types and CRDs. It will take place before validation. The sent object
+doesn't have to be valid, but fields will be cleared in order to make it valid.
+This process will also happen before fields tracking (server-side apply), so
+changes in discriminator, even if implicit, will be owned by the client making
+the update (and may result in conflicts).
+
+This process works as follows:
+- If there is a discriminator, and its value has changed, clear all fields but
+  the one specified by the discriminator,
+- If there is no discriminator, or if its value hasn't changed,
+  - if there is exactly one field set, set the discriminator (if exists) to that
```

Comment:
what does "if exists" mean here exactly? We fall through to the second step? or leave the discriminator missing?

Reply:
I'll clarify, thanks